cli/global_state.rs

```diff
@@ -141,13 +142,26 @@ impl GlobalState {
       .expect("Source file not found");
 
     // Check if we need to compile files
-    let needs_compilation = match out.media_type {
-      msg::MediaType::TypeScript
-      | msg::MediaType::TSX
-      | msg::MediaType::JSX => true,
-      msg::MediaType::JavaScript => self.ts_compiler.compile_js,
-      _ => false,
-    };
+    // Compilation happens if either:
+    // - `checkJs` is set to true in TS config
+    // - entry point is a TS file
+    // - any dependency in module graph is a TS file
+    let has_ts_dep = module_graph.values().any(|module_file| {
+      let media_type = module_file.media_type;
+
+      media_type == (MediaType::TypeScript as i32)
+        || media_type == (MediaType::TSX as i32)
+        || media_type == (MediaType::JSX as i32)
+    });
+
+    let needs_compilation = has_ts_dep
+      || (match out.media_type {
+        msg::MediaType::TypeScript
+        | msg::MediaType::TSX
+        | msg::MediaType::JSX => true,
+        msg::MediaType::JavaScript => self.ts_compiler.compile_js,
+        _ => false,
+      });
```

Comment:
drive-by commenter here; if the module graph may be large, wouldn't you rather run the simple match first?

Reply:
Doesn't really matter as it's the cheapest operation in whole compilation stack, but I refactored it to better reflect the intention

Reply:
thanks for having taken time to listen to a passerby on a low consequence suggestion :)